workflow-core/src/main/java/com/squareup/workflow1/Tracing.kt

```diff
@@ -0,0 +1,85 @@
+package com.squareup.workflow1
+
+/**
+ * Tracing facade to install later.
+ *
+ * This follows the static service locator that Timber demonstrates.
+ * See [Timber][https://github.com/JakeWharton/timber/blob/trunk/timber/src/main/java/timber/log/Timber.kt].
+ */
+public class Tracing private constructor() {
+
+  public interface Tracer {
+    public fun beginIdSection(label: () -> WorkflowIdentifier)
+    public fun endIdSection(label: () -> WorkflowIdentifier)
+    public fun beginSection(label: () -> String)
+    public fun endSection(label: () -> String)
+  }
+
+  public companion object : Tracer {
+
+    /**
+     * Note that by default we have nothing installed so nothing gets run! Only install something
+     * in an app you want to profile.
+     */
+    @JvmStatic public fun install(tracer: Tracer) {
+      require(tracer != this) { "Don't install Tracing on itself." }
+      synchronized(tracers) {
+        tracers.add(tracer)
+        tracerArray = tracers.toTypedArray()
+      }
+    }
+
+    @JvmStatic public fun uninstall(tracer: Tracer) {
+      synchronized(tracers) {
+        require(tracers.remove(tracer)) { "Can't uninstall tracer not installed." }
+        tracerArray = tracers.toTypedArray()
+      }
+    }
+
+    @JvmStatic public fun clear() {
+      synchronized(tracers) {
+        tracers.clear()
+        tracerArray = emptyArray()
+      }
+    }
+
+    public inline fun <T> traceId(
+      noinline label: () -> WorkflowIdentifier,
+      crossinline block: () -> T
+    ): T {
+      beginIdSection(label)
+      val result = block()
+      endIdSection(label)
+      return result
+    }
+
+    override fun beginIdSection(label: () -> WorkflowIdentifier) {
+      tracerArray.forEach { it.beginIdSection(label) }
+    }
+
+    override fun endIdSection(label: () -> WorkflowIdentifier) {
+      tracerArray.forEach { it.endIdSection(label) }
+    }
+
+    public inline fun <T> trace(
+      noinline label: () -> String,
+      crossinline block: () -> T
+    ): T {
+      beginSection(label)
+      val result = block()
+      endSection(label)
+      return result
+    }
+
+    override fun beginSection(label: () -> String) {
+      tracerArray.forEach { it.beginSection(label) }
+    }
+
+    override fun endSection(label: () -> String) {
+      tracerArray.forEach { it.endSection(label) }
+    }
+
+    private val tracers = ArrayList<Tracer>()
```

Comment:
note: YAGNI? Do you really need to support N Tracer impls, vs just the one (or none) ?

Reply:
ya its a good point. I think one is probably fine.